front/migrations/20250709_agent_mcp_action_fk_step_content.ts

```diff
@@ -0,0 +1,158 @@
+import { Op } from "sequelize";
+
+import { getWorkspaceInfos } from "@app/lib/api/workspace";
+import { AgentMCPAction } from "@app/lib/models/assistant/actions/mcp";
+import { AgentStepContentModel } from "@app/lib/models/assistant/agent_step_content";
+import { concurrentExecutor } from "@app/lib/utils/async_utils";
+import type { Logger } from "@app/logger/logger";
+import { makeScript } from "@app/scripts/helpers";
+import { runOnAllWorkspaces } from "@app/scripts/workspace_helpers";
+import type { LightWorkspaceType } from "@app/types";
+import { isFunctionCallContent } from "@app/types/assistant/agent_message_content";
+
+const FETCH_ACTIONS_BATCH_SIZE = 200;
+const WORKSPACE_CONCURRENCY = 5;
+const UPDATE_ACTION_CONCURRENCY = 10;
+
+/**
+ * Attach an AgentStepContent to all AgentMCPActions that don't have one for a given workspace.
+ *
+ * If no matching step content is found we skip the action. We do not warn or error log, unprocessed actions will be seen in DB.
+ * We only log at the end of the loop the number of actions skipped, updated and processed.
+ */
+async function attachStepContentToMcpActions({
+  workspace,
+  execute,
+}: {
+  workspace: LightWorkspaceType;
+  execute: boolean;
+}): Promise<{
+  workspaceId: string;
+  nbActionsToProcess: number;
+  nbActionsLinked: number;
+  nbActionsSkipped: number;
+}> {
+  let nbActionsToProcess = 0;
+  let nbActionsLinked = 0;
+  let nbActionsSkipped = 0;
+  let hasMore = false;
+  let lastId = 0;
+
+  do {
+    const mcpActions = await AgentMCPAction.findAll({
+      where: {
+        workspaceId: workspace.id,
+        id: {
+          [Op.gt]: lastId,
+        },
+        stepContentId: {
+          [Op.is]: null,
+        },
+      },
+      limit: FETCH_ACTIONS_BATCH_SIZE,
+      order: [["id", "ASC"]],
+    });
+
+    if (mcpActions.length === 0) {
+      return {
+        workspaceId: workspace.sId,
+        nbActionsToProcess: 0,
+        nbActionsLinked: 0,
+        nbActionsSkipped: 0,
+      };
+    }
+
+    const result = await concurrentExecutor(
+      mcpActions,
+      async (mcpAction) => {
+        // We look for a matching agent step content.
+        const stepContents = await AgentStepContentModel.findAll({
```

Comment:
I think there should be only one, @fontanierh do you confirm and then do you prefer if I put a findOne?